File: ios/Classes/Core/NativeStorageCore.swift
import Foundation

public class NativeStorageCore {
    public let spool: FileSpool

    public init(appGroupId: String? = nil) {
        let containerUrl: URL
        if let appGroupId = appGroupId, let groupUrl = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId) {
            containerUrl = groupUrl
        } else {
            containerUrl = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        }
        self.spool = FileSpool(containerUrl: containerUrl)
    }

    public func getPendingDeliveries() -> [NativeDelivery] {
        return spool.getPendingDeliveries().map { toNativeDelivery($0) }
    }

    public func claimDelivery(deliveryId: String, leaseDurationSeconds: Int64) throws -> NativeClaim {
        let claim = try spool.claimDelivery(deliveryId: deliveryId, leaseDurationSeconds: Double(leaseDurationSeconds))
        return NativeClaim(
            claimId: claim.claimId,
            deliveryId: claim.deliveryId,
            delivery: toNativeDelivery(claim.delivery),
            claimedAtUtc: claim.claimedAtUtc,
            expiresAtUtc: claim.expiresAtUtc
        )
    }

    public func claimNextDelivery(leaseDurationSeconds: Int64) throws -> NativeClaim? {
        let pending = spool.getPendingDeliveries()
        guard let first = pending.first else { return nil }
        return try claimDelivery(deliveryId: first.deliveryId, leaseDurationSeconds: leaseDurationSeconds)
    }

    public func acknowledgeClaim(claimId: String, deliveryId: String) throws {
        try spool.acknowledgeClaim(claimId: claimId, deliveryId: deliveryId)
    }

    public func releaseClaim(claimId: String, deliveryId: String, reason: String?) throws {
        try spool.releaseClaim(claimId: claimId, deliveryId: deliveryId, reason: reason)
    }

    public func retryDelivery(deliveryId: String) {
        // Automatically retried when lease expires
    }

    public func inspectInbox() -> NativeHealth {
        let pending = spool.getPendingDeliveries()
        return NativeHealth(
            pendingCount: Int64(pending.count),
            claimedCount: 0,
            quarantinedCount: 0,
            totalStorageBytes: 0,
            issues: []
        )
    }

    public func cleanupInbox(maxAgeSeconds: Int64) -> NativeCleanupResult {
        let (count, bytes) = spool.cleanup(maxAgeSeconds: Double(maxAgeSeconds))
        return NativeCleanupResult(
            deletedDeliveriesCount: Int64(count),
            reclaimedBytes: bytes
        )
    }

    public func getPayloadPath(deliveryId: String, itemId: String) throws -> String {
        return try spool.getPayloadUrl(deliveryId: deliveryId, itemId: itemId).path
    }

    private func toNativeDelivery(_ d: SpoolDelivery) -> NativeDelivery {
        return NativeDelivery(
            deliveryId: d.deliveryId,
            receivedAtUtc: d.receivedAtUtc,
            platform: d.platform,
            state: d.state,
            attempt: Int64(d.attempt),
            items: d.items.map { toNativeItem($0) },
            text: d.text,
            subject: d.subject,
            source: d.source
        )
    }

    private func toNativeItem(_ i: SpoolItem) -> NativeItem {
        return NativeItem(
            itemId: i.itemId,
            kind: i.kind,
            originalName: i.originalName,
            internalName: i.internalName,
            declaredMimeType: i.declaredMimeType,
            resolvedMimeType: i.resolvedMimeType,
            byteLength: i.byteLength
        )
    }
}

File: android/src/main/kotlin/dev/rayan/share_harbor/core/NativeStorageCore.kt
package dev.rayan.share_harbor.core

import android.content.Context
import dev.rayan.share_harbor.generated.NativeClaim
import dev.rayan.share_harbor.generated.NativeCleanupResult
import dev.rayan.share_harbor.generated.NativeDelivery
import dev.rayan.share_harbor.generated.NativeHealth
import dev.rayan.share_harbor.generated.NativeItem
import java.io.File

class NativeStorageCore(context: Context) {

    val spool = FileSpool(context.filesDir)

    fun getPendingDeliveries(): List<NativeDelivery> {
        val pending = spool.getPendingDeliveries()
        return pending.map { toNativeDelivery(it) }
    }

    fun claimDelivery(deliveryId: String, leaseDurationSeconds: Long): NativeClaim {
        val claim = spool.claimDelivery(deliveryId, leaseDurationSeconds)
        return NativeClaim(
            claimId = claim.claimId,
            deliveryId = claim.deliveryId,
            delivery = toNativeDelivery(claim.delivery),
            claimedAtUtc = claim.claimedAtUtc,
            expiresAtUtc = claim.expiresAtUtc
        )
    }

    fun claimNextDelivery(leaseDurationSeconds: Long): NativeClaim? {
        val pending = spool.getPendingDeliveries()
        if (pending.isEmpty()) return null
        return claimDelivery(pending.first().deliveryId, leaseDurationSeconds)
    }

    fun acknowledgeClaim(claimId: String, deliveryId: String) {
        spool.acknowledgeClaim(claimId, deliveryId)
    }

    fun releaseClaim(claimId: String, deliveryId: String, reason: String?) {
        spool.releaseClaim(claimId, deliveryId, reason)
    }

    fun retryDelivery(deliveryId: String) {
    }

    fun inspectInbox(): NativeHealth {
        val pending = spool.getPendingDeliveries()
        val spoolDir = File(contextFilesDir(spool), "share_harbor/v1/deliveries")
        var totalBytes = 0L

        if (spoolDir.exists()) {
            totalBytes = spoolDir.walkTopDown().sumOf { it.length() }
        }

        return NativeHealth(
            pendingCount = pending.size.toLong(),
            claimedCount = 0L,
            quarantinedCount = 0L,
            totalStorageBytes = totalBytes,
            issues = emptyList()
        )
    }

    fun cleanupInbox(maxAgeSeconds: Long): NativeCleanupResult {
        val (deletedCount, reclaimedBytes) = spool.cleanup(maxAgeSeconds)
        return NativeCleanupResult(
            deletedDeliveriesCount = deletedCount.toLong(),
            reclaimedBytes = reclaimedBytes
        )
    }

    fun getPayloadPath(deliveryId: String, itemId: String): String {
        return spool.getPayloadFile(deliveryId, itemId).absolutePath
    }

    private fun contextFilesDir(spool: FileSpool): File {
        return spool.getPendingDeliveries().let {
            File(System.getProperty("user.dir") ?: "/")
        }
    }

    private fun toNativeDelivery(d: SpoolDelivery): NativeDelivery {
        return NativeDelivery(
            deliveryId = d.deliveryId,
            receivedAtUtc = d.receivedAtUtc,
            platform = d.platform,
            state = d.state,
            attempt = d.attempt.toLong(),
            items = d.items.map { toNativeItem(it) },
            text = d.text,
            subject = d.subject,
            source = d.source
        )
    }

    private fun toNativeItem(i: SpoolItem): NativeItem {
        return NativeItem(
            itemId = i.itemId,
            kind = i.kind,
            originalName = i.originalName,
            internalName = i.internalName,
            declaredMimeType = i.declaredMimeType,
            resolvedMimeType = i.resolvedMimeType,
            byteLength = i.byteLength
        )
    }
}
